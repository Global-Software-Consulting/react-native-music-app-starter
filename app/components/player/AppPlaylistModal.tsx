import React from 'react';
import { View, TouchableOpacity, Button, FlatList } from 'react-native';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import { useTheme, Text } from 'react-native-paper';
import useStyles from './styles';
import Modal from 'react-native-modal';
import AppHeader from '../AppHeader';
import PlaylistSongsCard from '../Playlist/PlaylistSongs/PlaylistSongsCard';
import { PlayerState } from '../../models/reducers/index';
import { useDispatch, useSelector } from 'react-redux';
import { updatePlayList, deletePlayListFolder } from '../../store/actions/playerActions';
import Toast from 'react-native-simple-toast';

interface MusicProps {
    isModalVisible?: any;
    onPressPlaylist?: any;
    onPressNewPlaylist?: any;
}
interface IPState {
    playerReducer: PlayerState;
}
const AppPlaylistModal: React.FC<MusicProps> = ({
    isModalVisible,
    onPressPlaylist,
    onPressNewPlaylist,
}) => {
    const styles = useStyles();
    const theme = useTheme();
    const dispatch = useDispatch();
    const playList = useSelector((state: IPState) => state.playerReducer.playList);
    const selectedTrack: any = useSelector((state: IPState) => state.playerReducer.playerList);

    const addSongToPlaylist = (item: any) => {
        Toast.show(`Added in ${item.name} Playlist`);
        const found = item.songs.find((el: any) => el.id === selectedTrack.id);
        const data = item;
        if (!found) data.songs.push(selectedTrack);
        const list = playList.map((item: any) => {
            if (item.name === data.name) {
                return data;
            } else {
                return item;
            }
        });

        dispatch(updatePlayList(list));
    };

    const removePlaylist = (name: any) => {
        const data = playList?.filter((element: any) => element.name !== name);
        dispatch(deletePlayListFolder(data));
    };

    const PlayListRenderItem = ({ item }: any) => (
        <>
            <TouchableOpacity key={item} onPress={() => addSongToPlaylist(item)}>
                <PlaylistSongsCard
                    name={item.name}
                    model={item?.songs?.length}
                    img={
                        item.songs.length > 0
                            ? item.songs[0].artwork
                            : `https://picsum.photos/150/200/?random=${Math.random()}`
                    }
                    onPressRemove={removePlaylist}
                    showDel={true}
                    addSongToPlaylist={addSongToPlaylist}
                />
            </TouchableOpacity>
        </>
    );

    return (
        <>
            <View style={styles.container}>
                <Modal isVisible={isModalVisible}>
                    <View style={styles.ModalContainer}>
                        <AppHeader renderLeft={<Text style={styles.label}>Add To Playlist</Text>} />

                        <TouchableOpacity
                            style={styles.newListLabel}
                            onPress={() => onPressNewPlaylist()}>
                            <Text style={styles.name}>Create new playlist</Text>
                            <View style={{ width: '30%' }} />
                            <TouchableOpacity onPress={() => onPressNewPlaylist()}>
                                <MaterialCommunityIcons
                                    name="playlist-plus"
                                    style={[styles.secondaryControl, styles.off]}
                                    size={30}
                                    color={theme.colors.primary}
                                />
                            </TouchableOpacity>
                        </TouchableOpacity>
                        <View style={styles.folderContainer}>
                            {playList?.length > 0 ? (
                                <FlatList
                                    contentContainerStyle={{ alignSelf: 'flex-start' }}
                                    showsVerticalScrollIndicator={false}
                                    showsHorizontalScrollIndicator={false}
                                    data={playList}
                                    renderItem={PlayListRenderItem}
                                />
                            ) : (
                                <View style={styles.container}>
                                    <Text style={styles.model}>Playlist Empty</Text>
                                </View>
                            )}
                        </View>
                    </View>
                    <Button title="Cancel" onPress={() => onPressPlaylist()} />
                </Modal>
            </View>
        </>
    );
};

export default AppPlaylistModal;
