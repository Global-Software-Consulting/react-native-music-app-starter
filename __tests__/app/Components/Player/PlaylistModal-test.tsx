/**
 * @format
 */
import 'react-native';
import React from 'react';
import AppPlaylistModal from '../../../../app/components/player/AppPlaylistModal';
// Note: test renderer must be required after react-native.
import { render, fireEvent } from '@testing-library/react-native';


jest.mock("react-native-responsive-screen", () => {
    return {
        heightPercentageToDP: jest.fn(),
        widthPercentageToDP: jest.fn(),

    }
});

jest.mock('@react-navigation/native', () => {
    return {
        useIsFocused: jest.fn(),
        useNavigation: jest.fn(),
        // useRoute: jest.fn(),
        useRoute: () => {
            return {
                route: {
                    params: {
                        item: {
                            id: '5',
                            url: 'https://www.chosic.com/wp-content/uploads/2021/05/inossi-got-you.mp3',
                            title: 'Got You',
                            artist: 'Noor',
                            artwork: 'https://picsum.photos/id/103/200/300',
                            album: 'INOSSI',
                        }

                    }
                },
            }

        }

    }


});
jest.mock('react-i18next', () => {
    return {
        useTranslation: () => {
            return {
                t: jest.fn(),
            };
        },
    };
});

jest.mock('react-native-iphone-x-helper', () => ({
    getStatusBarHeight: jest.fn(),
    getBottomSpace: jest.fn(),
}));
jest.mock('react-redux', () => ({
    useDispatch: jest.fn(),
    useSelector: jest.fn(),
}));
// jest.mock('react-native-modal', () => 'react-native-modal');
jest.mock('react-native-modal', () => {
    return {
        __esModule: true,
        A: true,
        namedExport: jest.fn(),
        default: 'mockedDefaultExport',
    };
});

jest.mock('react-native-simple-toast', () => {
    return {
        Toast: jest.fn(),
    };
});
describe("group tesing goes here", () => {
    // test('change text', () => {
    //     const { getByTestId } = render(<AppPlaylistModal />);
    //     // fireEvent.press(getByTestId('changeTest'));
    //     // fireEvent(getByPlaceholderText('changeTest'), 'onChangeText', 'ab');
    //     fireEvent.changeText(getByTestId('changeTest'));
    // });
    test('pressing playlist modal', () => {
        const { getByTestId } = render(<AppPlaylistModal />);
        fireEvent.press(getByTestId('pressingPlaylist'));
    });
    test('new playlist modal', () => {
        const { getByTestId } = render(<AppPlaylistModal />);
        fireEvent.press(getByTestId('newlist'));
    });
    test('new playlist modal', () => {
        const { getByTestId } = render(<AppPlaylistModal />);
        fireEvent.press(getByTestId('createNewlist'));
    });
    test('new playlist modal', () => {
        const { getByTestId } = render(<AppPlaylistModal addSongToPlaylist={() => { }} />);
        fireEvent.press(getByTestId('newSongsAdded'));
    });
})

