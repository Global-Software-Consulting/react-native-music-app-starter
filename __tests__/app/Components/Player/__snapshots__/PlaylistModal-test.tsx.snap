// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`group tesing goes here renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#03dac4",
      "justifyContent": "center",
      "paddingLeft": 24,
      "paddingRight": 24,
    }
  }
>
  <react-native-modal>
    <View
      style={
        Object {
          "alignSelf": "center",
          "backgroundColor": "#03dac4",
          "borderRadius": 10,
          "height": "70%",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#03dac4",
            "borderBottomColor": "lightgray",
            "elevation": 5,
            "flexDirection": "row",
            "height": undefined,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.25,
            "shadowRadius": 3.84,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "padding": 7,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Object {
                  "fontSize": undefined,
                  "fontWeight": "bold",
                  "marginBottom": 12,
                  "paddingLeft": "5%",
                },
              ]
            }
          >
            Add To Playlist
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
                "width": "75%",
              }
            }
          />
          <View
            accessibilityRole="button"
            accessibilityState={Object {}}
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {},
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#007AFF",
                      "fontSize": 18,
                      "margin": 8,
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Cancel
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "alignSelf": "center",
              "flex": 1,
              "justifyContent": "center",
              "marginRight": 25,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "center",
                "marginRight": 5,
              }
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "fontSize": undefined,
                    "fontWeight": "bold",
                  },
                ]
              }
            />
          </View>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flexDirection": "row",
            "opacity": 1,
            "paddingTop": "5%",
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Object {
                "fontSize": undefined,
                "fontWeight": "bold",
                "paddingLeft": "5%",
              },
            ]
          }
        >
          Create new playlist
        </Text>
        <View
          style={
            Object {
              "justifyContent": "flex-end",
              "width": "42%",
            }
          }
        />
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <mockedDefaultExport
            color="#6200ee"
            name="playlist-plus"
            size={30}
            style={
              Array [
                Object {
                  "height": undefined,
                  "width": undefined,
                },
                Object {
                  "color": "#6200ee",
                  "opacity": 0.3,
                },
              ]
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#03dac4",
            "justifyContent": "center",
            "paddingLeft": 24,
            "paddingRight": 24,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "darkgray",
                "fontSize": undefined,
                "justifyContent": "center",
                "marginTop": -7,
                "textAlign": "center",
              },
            ]
          }
        >
          Playlist Empty
        </Text>
      </View>
    </View>
  </react-native-modal>
</View>
`;
